File: src/nac/token/token.cpp
#include "sodium/nac/token/token.h"

#include <string>

#include "sodium/nac/basic/source_range.h"

namespace sodium {

Token::Token(TokenKind kind, SourceRange range, uint32_t length) : kind_(kind), range_(range), length_(length) {}

TokenKind Token::kind() const {
    return kind_;
}

SourceRange Token::range() const {
    return range_;
}

uint32_t Token::length() const {
    return length_;
}

std::string Token::value() const {
    return std::string(range_.start().position(), length_);
}

Token Token::dummy() {
    return Token(TokenKind::EOF_TOKEN, SourceRange(), 0);
}

} // namespace sodium

File: nac/src/token/token.rs
use crate::basic::source_range::SourceRange;

#[derive(Debug, PartialEq, Eq)]
pub enum TokenKind {
    // reserved words
    Keyword(Keyword),
    Type(Type),

    // literal tokens
    Identifier(String),
    IntegerLiteral(i32),

    // symbols
    Arrow,
    LeftBrace,
    RightBrace,
    LeftParen,
    RightParen,
    Semicolon,

    // misc
    EOF,
    Unknown,
}

#[derive(Debug, PartialEq, Eq)]
pub enum Keyword {
    Func,
    Return,
}

#[derive(Debug, PartialEq, Eq)]
pub enum Type {
    Int,
}

pub struct Token {
    kind: TokenKind,
    range: SourceRange,
}

impl Token {
    pub fn new(kind: TokenKind, range: SourceRange) -> Self {
        Self { kind, range }
    }

    pub fn kind(&self) -> &TokenKind {
        &self.kind
    }

    pub fn range(&self) -> &SourceRange {
        &self.range
    }
}
